components/director/internal/systemfetcher/client.go

```diff
@@ -0,0 +1,125 @@
+package systemfetcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	urlpkg "net/url"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+var (
+	scopes = []string{"uaa.resource"}
+)
+
+type OAuth2Config struct {
+	ClientID                  string `envconfig:"APP_OAUTH_CLIENT_ID"`
+	ClientSecret              string `envconfig:"APP_OAUTH_CLIENT_SECRET"`
+	OAuthTokenEndpointPattern string `envconfig:"APP_OAUTH_TOKEN_ENDPOINT_PATTERN"`
+}
+
+type APIConfig struct {
+	Endpoint                    string        `envconfig:"APP_SYSTEM_INFORMATION_ENDPOINT"`
+	FilterCriteria              string        `envconfig:"APP_SYSTEM_INFORMATION_FILTER_CRITERIA"`
+	FilterTenantCriteriaPattern string        `envconfig:"APP_SYSTEM_INFORMATION_FILTER_TENANT_CRITERIA_PATTERN"`
+	Timeout                     time.Duration `envconfig:"APP_SYSTEM_INFORMATION_FETCH_TIMEOUT"`
+}
+
+type Client struct {
+	apiConfig         APIConfig
+	oAuth2Config      OAuth2Config
+	clientCreatorFunc ClientCreatorFunc
+}
+
+type ClientCreatorFunc func(ctx context.Context, oauth2Config OAuth2Config, scopes []string, tenant string) *http.Client
+
+func DefaultClientCreator(ctx context.Context, oAuth2Config OAuth2Config, scopes []string, tenant string) *http.Client {
+	cfg := clientcredentials.Config{
+		ClientID:     oAuth2Config.ClientID,
+		ClientSecret: oAuth2Config.ClientSecret,
+		TokenURL:     oAuth2Config.OAuthTokenEndpointPattern,
+		Scopes:       scopes,
+	}
+
+	httpClient := cfg.Client(ctx)
+	return httpClient
+}
+
+func NewClient(apiConfig APIConfig, oAuth2Config OAuth2Config, clientCreatorFunc ClientCreatorFunc) *Client {
+	return &Client{
+		apiConfig:         apiConfig,
+		oAuth2Config:      oAuth2Config,
+		clientCreatorFunc: clientCreatorFunc,
+	}
+}
+
+func (c *Client) FetchSystemsForTenant(ctx context.Context, tenant string) ([]System, error) {
```

Comment:
Can you add a brief comment on this function about what it does and particularly why it seemingly makes the same HTTP remote call twice (but with different query params)?